### PR TITLE
Versions are Unique to Applications

### PR DESCRIPTION
**Pull request by larkmullins**

### CODE REVIEW

File: database/migrations/8_removeVersionsToEnvironmentsConstraint.go [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+import (
+	"github.com/rubenv/sql-migrate"
+)
+
+func removeVersionsToEnvironmentsConstraint() *migrate.Migration {
+	return &migrate.Migration{
+		Id: "8",
+		Up: []string{`
+      ALTER TABLE versions
```
Comment (jlaswell) [inline database/migrations/8_removeVersionsToEnvironmentsConstraint.go:11]: indentation

File: database/migrations/9_addApplicationUuidColumnToVersionsTable.go [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+import (
+	"github.com/rubenv/sql-migrate"
+)
+
+func addApplicationUuidColumnToVersionsTable() *migrate.Migration {
+	return &migrate.Migration{
+		Id: "9",
+		Up: []string{`
+      ALTER TABLE versions
```
Comment (jlaswell) [inline database/migrations/9_addApplicationUuidColumnToVersionsTable.go]: indentation

File: resources/version.go [hunk 3]
```diff
@@ -47,17 +47,20 @@ type ContainerVolumeMount struct {
...
-	Name          string      `gorm:"not null;unique_index:uix_versions_name_environment_id" json:"name"`
-	Slug          string      `gorm:"not null" json:"slug"`
-	Replicas      int         `gorm:"not null" sql:"DEFAULT:'0'" json:"replicas"`
-	EnvironmentID uint        `gorm:"not null;unique_index:uix_versions_name_environment_id" json:"environment_id"`
-	Environment   Environment `gorm:"not null" json:"environment"`
-	Volumes       []Volume    `gorm:"-" json:"volumes"`
-	Containers    []Container `gorm:"-" json:"containers"`
-	Spec          string      `gorm:"type:json" json:"-"`
+	UUID            string       `gorm:"not null;unique_index:uix_versions_uuid" sql:"type:varchar(36)" json:"uuid"`
+	Name            string       `gorm:"not null" json:"name"`
```
Comment (jlaswell) [inline resources/version.go]: Needs the `unique_index:uix_versions_name_application_uuid` listed w/in the field tag.

File: resources/version.go [hunk 4]
```diff
@@ -103,7 +106,7 @@ func CreateVersion(db *gorm.DB, client kube.APIInterface, version *Version) (boo
 
 	// update environment service
 	ports := gatherContainerPorts(version.Containers)
-	UpdateEnvironmentService(db, client, &version.Environment, ports)
+	UpdateEnvironmentService(db, client, version.Environment, ports)
```
Comment (jlaswell) [inline resources/version.go:122]: Any specific reason to change this from a reference?

**Review by jlaswell — DISMISSED** [review]
Somethings to note, but this is taking shape! 🎉

File: resources/version.go (hunk 4 above)
Comment (larkmullins) [reply resources/version.go:122]: There was an error regarding recursive type on the Version struct that was related to the Environment field type. To fix this I needed to set the Environment field on the Version struct to a pointer.

Comment (jlaswell) [reply resources/version.go:122]: Ah, I've seen that issue. Makes sense :+1:

File: database/migrations/3_createEnvironmentsTable.go [hunk 5]
```diff
@@ -16,6 +16,9 @@ func createEnvironmentsTable() *migrate.Migration {
         name varchar(255) NOT NULL,
         slug varchar(255) NOT NULL,
         application_id bigint(20) unsigned DEFAULT NULL,
+        application_uuid varchar(36) NOT NULL,
```
Comment (jlaswell) [inline database/migrations/3_createEnvironmentsTable.go]: We can't make changes to existing migrations now that we have deployed. Every change will need to happen in a newly created migration.

File: database/migrations/4_createVersionsTable.go [hunk 6]
```diff
@@ -16,11 +16,14 @@ func createVersionsTable() *migrate.Migration {
         name varchar(255) NOT NULL,
         slug varchar(255) NOT NULL,
         replicas int(11) NOT NULL DEFAULT '0',
+        application_id int(10) unsigned NOT NULL,
```
Comment (jlaswell) [inline database/migrations/4_createVersionsTable.go]: Existing migration changes.

File: database/migrations/8_removeVersionsToEnvironmentsConstraint.go [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+
+import (
+	"github.com/rubenv/sql-migrate"
+)
+
+func removeVersionsToEnvironmentsConstraint() *migrate.Migration {
+	return &migrate.Migration{
+		Id: "8",
+		Up: []string{`
+	      ALTER TABLE versions
```
Comment (jlaswell) [inline database/migrations/8_removeVersionsToEnvironmentsConstraint.go]: Indentation issue seems to be present still.

File: database/migrations/9_addApplicationUuidColumnToVersionsTable.go [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+
+import (
+	"github.com/rubenv/sql-migrate"
+)
+
+func addApplicationUUIDColumnToVersionsTable() *migrate.Migration {
+	return &migrate.Migration{
+		Id: "9",
+		Up: []string{`
+          ALTER TABLE versions
```
Comment (jlaswell) [inline database/migrations/9_addApplicationUuidColumnToVersionsTable.go]: \t

File: resources/environment.go [hunk 9]
```diff
@@ -16,12 +16,16 @@ import (
...
+	UUID            string      `gorm:"not null;unique_index" sql:"type:varchar(36)" json:"uuid"`
+	Name            string      `gorm:"not null" json:"name"`
+	Slug            string      `gorm:"not null" json:"slug"`
+	VersionID       uint64      `gorm:"not null" json:"version_id,string"`
+	VersionUUID     string      `gorm:"not null" json:"version_uuid"`
+	Version         Version     `json:"version"`
+	ApplicationID   uint64      `gorm:"not null" json:"application_id,string"`
+	ApplicationUUID string      `gorm:"not null" json:"application_uuid"`
+	Application     Application `json:"application"`
+	//Versions        []Version   `json:"versions,array,omitempty"`
```
Comment (jlaswell) [inline resources/environment.go]: We should remove this since it's just leftovers.

File: resources/version_test.go [hunk 10]
```diff
@@ -17,24 +17,31 @@ import (
...
-	mockClient.On("CreateOrUpdateDeployment", mock.Anything).Return(expectedError)
-
-	db, _ := gorm.Open("testdb", "")
-	result, err := CreateVersion(db, mockClient, version)
-	assert.False(t, result)
-	assert.Equal(t, "foo error", err.Error())
-
-	mockClient.AssertExpectations(t)
+	fmt.Println(version)
+	/*
```
Comment (jlaswell) [inline resources/version_test.go]: Can be removed.

File: resources/version_test.go [hunk 11]
```diff
@@ -81,6 +89,8 @@ func TestVersionDeploymentDefinition(t *testing.T) {
 	}
 
 	deployment := versionDeploymentDefinition(version)
+	// fmt.Println("+++HERE+++")
```
Comment (jlaswell) [inline resources/version_test.go]: Debug leftovers.

File: app/handlers/api/environments.go [hunk 12]
```diff
@@ -35,11 +35,13 @@ func EnvironmentIndex(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	for i := range environments {
-		if len(environments[i].Versions) == 0 {
-			environments[i].Versions = make([]resources.Version, 0)
+	/*
```
Comment (jlaswell) [inline app/handlers/api/environments.go]: Leftovers.

File: app/handlers/api/environments.go [hunk 13]
```diff
@@ -69,9 +71,11 @@ func EnvironmentShow(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	if len(environment.Versions) == 0 {
-		environment.Versions = make([]resources.Version, 0)
-	}
+	/*
```
Comment (jlaswell) [inline app/handlers/api/environments.go]: Leftovers.

**Review by jlaswell — DISMISSED** [review]
This has turned out to be way more involved than we thought originally for sure! Thanks for tackling this.

There are some leftovers and migration issues to address before we can merge.

File: database/migrations/3_createEnvironmentsTable.go (hunk 5 above)
Comment (larkmullins) [reply database/migrations/3_createEnvironmentsTable.go]: Thought was to add this in for future Brizo users but you are correct, newly created migrations would be correct.

File: database/migrations/4_createVersionsTable.go (hunk 6 above)
Comment (larkmullins) [reply database/migrations/4_createVersionsTable.go]: See previous comment

**Pull request merged**